File: 22-Binary-Trees-Miscellaneous/python/04_distinct_numbers_in_each_subarray.py
from collections import defaultdict
from typing import List

class Solution:
    def distinctNumbers(self, nums: List[int], k: int) -> List[int]:
        ans = []
        freq = defaultdict(int)
        n = len(nums)

        for i in range(k):
            freq[nums[i]] += 1
        ans.append(len(freq))

        for i in range(k, n):
            outgoing = nums[i - k]
            freq[outgoing] -= 1
            if freq[outgoing] == 0:
                del freq[outgoing]

            freq[nums[i]] += 1
            ans.append(len(freq))

        return ans

File: 22-Binary-Trees-Miscellaneous/cpp/04-distinct-numbers-in-each-subarray.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> distinctNumbers(vector<int>& nums, int k) {
        vector<int> ans;
        unordered_map<int, int> freq;
        int n = nums.size();

        for (int i = 0; i < k; i++) {
            freq[nums[i]]++;
        }
        ans.push_back(freq.size());

        for (int i = k; i < n; i++) {
            int outgoing = nums[i - k];
            freq[outgoing]--;
            if (freq[outgoing] == 0) {
                freq.erase(outgoing);
            }

            freq[nums[i]]++;
            ans.push_back(freq.size());
        }

        return ans;
    }
};
